Updates for Terraform policy docs (#2864)
hashicorp/web-unified-docs: content/terraform-policy/v0.1.x (beta)/docs/policy/reference/test/resource.mdx

## Configuration model

The `resource` block supports the following configuration:

- `resource` "[`<resource_type>`](#resource-type)" &nbsp "[`<test_label>`](#test-label)" &nbsp block
  - [`attrs`](#attrs) &nbsp object | optional
    - `<attribute_name>` &nbsp any | repeatable
  - [`expect_failure`](#expect-failure) &nbsp boolean | optional
  - [`inputs`](#inputs) &nbsp block | optional
    - `<input_name>` &nbsp any | repeatable
  - [`meta`](#meta) &nbsp object | optional
    - `module_path` &nbsp string
    - `operation` &nbsp string
    - `provider_type` &nbsp string
    - `type` &nbsp string
  - [`prior_attrs`](#prior-attrs) &nbsp object | optional
    - `<attribute_name>` &nbsp any | repeatable

>   - [`skip`](#skip) &nbsp boolean | optional

Reviewer: ```suggestion
    - [`<attribute_name>`](#attrs) &nbsp any | repeatable
  - [`expect_failure`](#expect_failure) &nbsp boolean | optional
  - [`inputs`](#inputs) &nbsp block | optional
    - [`<input_name>`](#inputs) &nbsp any | repeatable
  - [`meta`](#meta) &nbsp object | optional
    - [`module_path`](#meta) &nbsp string
    - [`operation`](#meta) &nbsp string
    - [`provider_type`](#meta) &nbsp string
    - [`type`](#meta) &nbsp string
  - [`prior_attrs`](#prior_attrs) &nbsp object | optional
    - [`<attribute_name>`](#prior_attrs) &nbsp any | repeatable
  - [`skip`](#skip) &nbsp boolean | optional
```

## Complete configuration example

> The following example demonstrates a resource block with all configuration options specified:

Reviewer: ```suggestion
The following example demonstrates a `resource` block with all configuration options specified:
```

      Environment = "production"
    }
  }

  inputs {
    max_instance_size = "t3.large"
  }
}
```

</CodeBlockConfig>

## Specification

A `resource` block supports the following configuration.

### Resource type

> The resource type (For example: `aws_instance`, `google_compute_instance`). Refer to the provider documentation for a list of supported resources.

Reviewer: ```suggestion
The resource type, such as `aws_instance` and `google_compute_instance`. Refer to the provider documentation for a list of supported resources.
```

| Operation | `attrs.<name>` |
|-----------|----------------------------|
| create    | Target State               |
| update    | Target State               |
| delete    | Target State (all resource attributes are null) |

### `expect_failure`

Set to `true` to expect this mock to fail the matching policies.

- Data type: Boolean
- Default: `false`

### `inputs`

Set policy input values specific to this test case.

- Data type: Block
- Default: None

### `meta`

Metadata for the resource, including its lifecycle operation.

- Data type: Object
- Default: None

You can set the following meta values in the `meta` argument.

| Attribute | Type | Description | Example |
|-----------|------|-------------|---------|
| `module_path` | String | The module path the resource belongs to. | `"."`, `"my-module"` |
| `operation` | String | The operation being performed. | `"create"`, `"update"`, `"delete"` |
| `provider_type` | String | The name of the resource provider. | `"aws"`, `"azurerm"` |
| `type` | String | The resource type. This is the first label of the matching resource block. | `"aws_s3_bucket"`, `"azurerm_managed_disk"` |

### `prior_attrs`

> The resource attributes before the operation (for `update` or `delete` operations). Each provider defines the value attributes for each of its resources. Refer to the provider documentation for more information.

Reviewer: ```suggestion
Specifies the resource attributes before the operation for `update` or `delete` operations. Each provider defines the value attributes for each of its resources. Refer to the provider documentation for more information.
```

| Block Type | Terraform policy attribute value |
|------------|----------------------------------|
| Single nested | Assign an object value to an attribute with the same name as the block. |
| List nested | Assign a list of object values. |
| Set nested  | Assign a set of object values. |

### Single nested block

> The following example demonstrates how to assign a value to the "admin_ssh_key" and "os_disk" single nested blocks.

Reviewer: ```suggestion
The following example demonstrates how to assign a value to the `admin_ssh_key` and `os_disk` single nested blocks.
```
Not a requirement, but we should consider moving these to the Examples section and linking from here so that the content structure is consistent.

Author: So for this particular construct, it seems really difficult for a reader (well, me anyway) to visualize what's going on without an inline example.

}
```

### List or set nested blocks

> The following example demonstrates how to assign set or list of values to the "service_account" nested block. The syntax for assigning set and list values is the same.

Reviewer: ```suggestion
The following example demonstrates how to assign set or list of values to the `service_account` nested block. The syntax for assigning set and list values is the same.
```

## Examples

The following examples demonstrate common mock resource configuration patterns for specific use cases.

### Test encryption requirement

> In the following example, two mock `resource` blocks test an encryption policy - one that will pass and one that will fail.

Reviewer: ```suggestion
In the following example, two mock `resource` blocks test an encryption policy. The first block mocks a passing scenario and the second block mocks the failing scenario.
```
